Add F5 Rancid Config Parser

jeffmccune opened the pull request:
This patch adds a command line tool to parse F5 configuration data
obtained from Rancid.  The purpose is to transform the F5 configuration,
which may contain many thousands of pools, virtual servers, and iRule
objects, into CSV output which may be easily analyzed using Google
Sheets.

Some select commits from the previous repository:

Add example to list currently active iRules.  Swati requested this
functionality working with the customer.

The examples also include the vip and destination in the outputs.

Add f5 print-irule subcommand

This patch adds a subcommand to print out an irule given a name.  The
purpose is to print out all iRules in a format easily processed.  The
irule can optionally be written to a file and also optionally in
markdown using `--format markdown`

Add jq example to group rules by target pool

This patch adds a jq example grouping all of the results by the target
pool.  The intent is to make it useful to focus on the rules applying to
an application service.

Add map subcommand to output a list of virtual servers

This is a JSON list of objects which can be filtered as per the [Filter
objects based on the contents of a
key](https://github.com/stedolan/jq/wiki/Cookbook#filter-objects-based-on-the-contents-of-a-key)
example in the jq Cookbook.

Add virtuals subcommand to output all virtual server names

This patch adds functionality to print out all virtual server names.
This builds toward mapping target pools to applicable iRules.

Add irules and pools subcommands

This patch adds an irules and pools subcommand to print all iRules given the
F5 configuration.  This establishes the basic structure for parsing the
F5 configuration and extracting information.

Client connections enter a Virtual Server.  iRules are bound to a
virtual server.  Connections are forwarded on to target pools.

jeffmccune commented:
I signed it!

jeffmccune commented:
I signed it!

morgante commented:
@aaron-lane to provide an initial review.

aaron-lane commented on tools/f5map/lib/f5map/parser.rb:
```diff
@@ -0,0 +1,34 @@
+  class Parser
+    # Load the grammar
+    Treetop.load(File.expand_path('../f5.treetop', __FILE__))
+    @@parser = F5PersistParser.new
```
`@@parser` does not appear to be used in any context other than `.parse` as `.parser is never called`. This should be inlined to `.parse` and removed from the class variables.

jeffmccune replied:
Ack

aaron-lane commented on tools/f5map/lib/f5map/parser.rb:
```diff
@@ -0,0 +1,34 @@
+
+    def self.parse(data)
+      tree = @@parser.parse(data)
+      # self.clean_tree!(tree)
```
Commented-out code should be removed.

aaron-lane commented on tools/f5map/lib/f5map/parser.rb:
```diff
@@ -0,0 +1,34 @@
+    private
+
+    # Would be nice to make this functional instead of imperative.
+    def self.clean_tree!(root_node)
```
This method is not called.

jeffmccune replied:
Will look into this...  That strikes me as strange.  I may have committed the code wired for debugging.

aaron-lane replied:
For reference, line 17 shows the only call to this method commented out.

jeffmccune commented on tools/f5map/f5map.gemspec:
```diff
@@ -0,0 +1,39 @@
+  spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
+  spec.require_paths = ["lib"]
+
+  spec.add_development_dependency "bundler", "~> 1.16"
```
Yes, and as a result it's a required development dependency...  ;) I think this is why it's automatically populated in the gemspec when you run `bundle gem foo`

aaron-lane replied:
That's odd. :grin:

jeffmccune commented on tools/f5map/README.md:
```diff
@@ -0,0 +1,330 @@
+
+With these pre-requisites the `f5` utility may be run from the command line or
+a CI system without installing anything else into the system paths.  All
+dependencies are intended to be self-contained in the workspace directory.
```
Gem dependencies, I'll clarify.  This was intended for the customer, they had concerns about installing ruby dependencies into system directory.  This is designed to be self-contained, e.g. `bundle install --path .bundle/gems/`

jeffmccune commented on tools/f5map/lib/f5map.rb:
```diff
@@ -0,0 +1,9 @@
+require 'f5map/version'
+
+module F5map
+  # Your code goes here...
```
Ack, will remove

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+  def reset!
+    reset_options!
+    reset_logging!(opts)
+    # Reverse Lookup Cache
```
Ack, will remove

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+  # @return [String, false] the code defining the named iRule or false if no irule is
+  #   found in the configuration.
+  def irule(name, config)
+    @config_str ||= input_stream(opts[:input]) {|io| io.read }
```
I'll look into it, but instance initialization happens really early, before config parsing I believe.

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+  def persist(virtual)
+    # Match to end to parse it out
+    re = /persist\s+(.*)/mx
+    rv = []
```
Ack, will change

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+    rv = []
+    if md = re.match(virtual.content)
+      if ast = Parser.parse(md.to_s.gsub("\n", ' '))
+        rv = ast.value[1].each_with_object([]) do |v, m|
```
FYI, I tend to prefer a functional style with explicit dependency injection, versus an OO style relying on the scope of the closure.  This is why you see me using `each_with_object` instead of `each`.

With that said, will change.

aaron-lane replied:
That's fair. Don't worry about it then.

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+        end
+      else
+        require 'pry'; binding.pry
+        rv
```
Yeah, and that debugger line shouldn't be there...

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+    if @reverse_lookup_cache[ip]
+      return @reverse_lookup_cache[ip]
+    end
+    fqdn_ary = Socket.getaddrinfo(ip, 0, Socket::AF_UNSPEC, Socket::SOCK_STREAM, nil, Socket::AI_CANONNAME, true).collect { |a| a[2]  }
```
I'd prefer to keep this as-is for readability.  I tripped over this, Socket returning an array, etc...  The comment and intermediate variable help with comprehension.  I'm also not sure if `.first[2]` is a typo, I'm a bit confused by the `[2]` index.

aaron-lane replied:
The present logic collects `[2]` for each element before selecting the `first` element, so they should be functionally equivalent.

jeffmccune commented on tools/f5map/lib/f5map/app.rb:
```diff
@@ -0,0 +1,336 @@
+  end
+
+  # regexp to match a destination entry in the `lvm virtual` content
+  DESTINATION_RE = %r{destination\s+(?<destination>(?<ip>[^\s]+):(?<protocol>[^\s+]+))}
```
Ack will change

jeffmccune commented on tools/f5map/lib/f5map/node_extensions.rb:
```diff
@@ -0,0 +1,74 @@
+module F5Persist
+  module Values
+    def values(elements)
+      elements.each_with_object([]) do |node, ary|
```
Ack

jeffmccune commented:
Thanks for the detailed review @aaron-lane I'll shoot to get as much of this completed by the end of the week.

morgante commented:
@jeffmccune @aaron-lane Any updates here?

jeffmccune commented:
Thanks for following up, this is on me but currently buried and digging
out.  Don't have an ETA on addressing the review.

May be good to close it and re-open time permitting?

morgante commented:
@jeffmccune Any chance you could take a stab at the review feedback?
